### PR TITLE
Implement max file size for mysql

### PR DESCRIPTION
**Pull request by bularond**
Add config WALG_STREAM_SPLITTER_MAX_FILE_SIZE witch set maximum one file size for backup

### CODE REVIEW

**Comment by ostinru** [comment]
@bularond , could you please elaborate on why do you want to limit file sizes? If I remember correctly, current implementation shouldn't keep whole files locally, so it is unclear what you are trying to improve here.

**Comment by mialinx** [comment]
> @bularond , could you please elaborate on why do you want to limit file sizes? If I remember correctly, current implementation shouldn't keep whole files locally, so it is unclear what you are trying to improve here.

Main idea behind `max size` is to avoid uploading / downloading huge files, as it consumes a lot of memory (s3client multipart specific)

File: internal/stream_push_helper.go [hunk 1]
```diff
@@ -39,13 +39,31 @@ func (uploader *SplitStreamUploader) PushStream(stream io.Reader) (string, error
...
-		dstPath := GetPartitionedStreamName(backupName, uploader.Compressor.FileExtension(), i)
-		tracelog.InfoLogger.Printf("Uploading... %v", dstPath)
-		errGroup.Go(func() error {
-			return uploader.PushStreamToDestination(reader, dstPath)
-		})
+	for partNumber := 0; partNumber < uploader.partitions; partNumber++ {
+		reader := readers[partNumber]
+		if uploader.maxFileSize != 0 {
+			tracelog.DebugLogger.Printf("Start backup part %d\n", partNumber)
+			filesReaders := splitmerge.SplitMaxSizeReader(reader, uploader.blockSize, uploader.maxFileSize)
```
Comment (mialinx) [inline internal/stream_push_helper.go:46]: Does all the `fileReaders` read from `reader` at the same time ?

File: internal/splitmerge/splitmaxsizereader.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"github.com/wal-g/tracelog"
+	"io"
+)
+
+func SplitMaxSizeReader(reader io.Reader, blockSize int, maxSize int) <-chan io.ReadCloser {
+	out := make(chan io.ReadCloser)
+
+	go func() {
```
Comment (mialinx) [inline internal/splitmerge/splitmaxsizereader.go:11]: why do we need a separate goroutine here ?

**Review by mialinx — CHANGES_REQUESTED** [review]
redesign needed

**Comment by ostinru** [comment]
FYI: we have PR that limits size of files for single-stream backups: https://github.com/wal-g/wal-g/pull/1013 (PR weirdly named).

File: docker/mysql_tests/scripts/base_tests/splited_file_stream.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
+set -e -x
+
+. /usr/local/export_common.sh
+
+export WALE_S3_PREFIX=s3://mysqlfullxtrabackupbucket
```
Comment (ostinru) [inline docker/mysql_tests/scripts/base_tests/splited_file_stream.sh:6]: Let's use a separate bucket for each test.
Don't forget to create it in `docker-compose.yml`

**Comment by mialinx** [comment]
> FYI: we have PR that limits size of files for single-stream backups: #1013 (PR weirdly named).

Yes, but... it does not support split/merge parallel processing.  And it stores redundant info (Filenames) in sentinel.

File: internal/stream_metadata.go [hunk 4]
```diff
@@ -32,12 +33,23 @@ func GetBackupStreamFetcher(backup Backup) (StreamFetcher, error) {
 		return nil, err
 	}
 
+	limitedFileSize := false
+	if metadata.Type == LimitedFileSizeBackup {
+		limitedFileSize = true
+	}
+
 	switch metadata.Type {
+	case LimitedFileSizeBackup:
```
Comment (mialinx) [inline internal/stream_metadata.go:42]: It looks like thees two cases differ only in the last argument. May be join them together ?  And get rid of temporary  `limitedFileSize` variable

File: internal/stream_fetch_helper.go [hunk 5]
```diff
@@ -132,11 +111,66 @@ func DownloadAndDecompressSplittedStream(backup Backup, blockSize int, extension
 	return lastErr
 }
 
-func detectPartitionsCount(backup Backup, decompressor compression.Decompressor) (int, error) {
+func downloadFile(backup Backup, decompressor compression.Decompressor, fileName string, writer io.WriteCloser) error {
```
Comment (mialinx) [inline internal/stream_fetch_helper.go:114]: We have added 4 extra functions.

May we just modify `detectPartitionsCount` to return `[]int` (number of part for each partition).
And add one loop (over parts) in `DownloadAndDecompressSplittedStream` ?

**Review by mialinx — CHANGES_REQUESTED** [review]

**Review by mialinx — CHANGES_REQUESTED** [review]

**Review by mialinx — APPROVED** [review]
LGTM

**Pull request merged**